pkg/bundle/manager.go

```diff
@@ -144,84 +127,88 @@ func (m bundleManager) SortBundlesDescending(bundles []api.PackageBundle) {
+		if pbc.Status.State == api.BundleControllerStateActive {
+			m.log.Error(err, "marking disconnected")
+			pbc.Status.State = api.BundleControllerStateDisconnected
+			err = m.bundleClient.SaveStatus(ctx, pbc)
+			if err != nil {
+				return fmt.Errorf("updating %s status to %s: %s", pbc.Name, pbc.Status.State, err)
+			}
+		}
+		return nil
 	}
 
-	return bundle, nil
-}
-
-func (m *bundleManager) ProcessLatestBundle(ctx context.Context, latestBundle *api.PackageBundle) error {
 	knownBundles := &api.PackageBundleList{}
-	err := m.bundleClient.GetBundleList(ctx, knownBundles)
+	err = m.bundleClient.GetBundleList(ctx, knownBundles)
 	if err != nil {
 		return fmt.Errorf("getting bundle list: %s", err)
 	}
 
+	found := false
+	latest := true
 	for _, b := range knownBundles.Items {
 		if b.Name == latestBundle.Name {
-			return nil
+			found = true
+			break
 		}
+		latest = false
 	}
 
-	err = m.bundleClient.CreateBundle(ctx, latestBundle)
-	if err != nil {
-		return fmt.Errorf("creating new package bundle: %s", err)
-	}
-
-	return nil
-}
-
-func kubeVersion(name string) (string, error) {
-	matches := kubeVersionRe.FindStringSubmatch(name)
-	if len(matches) > 1 {
-		return matches[1], nil
+	if !found {
+		err = m.bundleClient.CreateBundle(ctx, latestBundle)
+		if err != nil {
+			return fmt.Errorf("creating new package bundle: %s", err)
+		}
 	}
 
-	return "", fmt.Errorf("no kubernetes version found in %q", name)
-}
-
-var kubeVersionRe = regexp.MustCompile(`^(v[^-]+)-.*$`)
-
-func (m *bundleManager) apiVersion() (string, error) {
-	info, err := m.kubeServerVersion.ServerVersion()
-	if err != nil {
-		return "", fmt.Errorf("getting server version: %s", err)
+	switch pbc.Status.State {
+	case api.BundleControllerStateActive:
+		if latest {
+			break
+		}
+		pbc.Status.State = api.BundleControllerStateUpgradeAvailable
+		m.log.V(6).Info("update", "PackageBundleController", pbc.Name, "state", pbc.Status.State)
+		err = m.bundleClient.SaveStatus(ctx, pbc)
+		if err != nil {
+			return fmt.Errorf("updating %s status to %s: %s", pbc.Name, pbc.Status.State, err)
+		}
+	case api.BundleControllerStateUpgradeAvailable:
+		if !latest {
+			break
+		}
+		pbc.Status.State = api.BundleControllerStateActive
+		m.log.V(6).Info("update", "PackageBundleController", pbc.Name, "state", pbc.Status.State)
+		err = m.bundleClient.SaveStatus(ctx, pbc)
+		if err != nil {
+			return fmt.Errorf("updating %s status to %s: %s", pbc.Name, pbc.Status.State, err)
+		}
+	case api.BundleControllerStateDisconnected:
+		pbc.Status.State = api.BundleControllerStateActive
+		m.log.V(6).Info("update", "PackageBundleController", pbc.Name, "state", pbc.Status.State)
+		err = m.bundleClient.SaveStatus(ctx, pbc)
+		if err != nil {
+			return fmt.Errorf("updating %s status to %s: %s", pbc.Name, pbc.Status.State, err)
+		}
+	case "":
+		if pbc.Spec.ActiveBundle == "" {
+			pbc.Spec.ActiveBundle = latestBundle.Name
+		}
+		pbc.Status.State = api.BundleControllerStateActive
```

Comment:
Does this change get applied?

> The PUT and POST verbs on objects MUST ignore the status values, to avoid accidentally overwriting the status in read-modify-write scenarios. A /status subresource MUST be provided to enable system components to update statuses of resources they manage.
> -- https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status

It's possible this isn't having an effect? Though I guess, it'll get picked up at the next reconcile cycle anyway? But this might need it's own specific SaveStatus call?

Reply:
Done. Three reconcile loops caused by this, I don't think it could be done in less.
```
 % k get -n eksa-packages packages,packagebundles,packagebundlecontroller
NAME                                                STATE
packagebundle.packages.eks.amazonaws.com/v1-21-33   active

NAME                                                                                 ACTIVEBUNDLE   STATE    DETAIL
packagebundlecontroller.packages.eks.amazonaws.com/eksa-packages-bundle-controller   v1-21-33       active
```